### PR TITLE
test: watch full stdout when waiting for match

### PR DESCRIPTION
**Pull request by jbedard**
... and cleanup a few typescript types.

I've been debugging a failing test where stderr is not being matched. I tried these fixes and figured they may as well be merged? 🤷

### CODE REVIEW

File: tests/legacy-cli/e2e/tests/build/bundle-budgets.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export default async function () {
     ];
   });
 
-  const errorMessage = await expectToFail(() => ng('build'));
```
Comment (jbedard) [inline tests/legacy-cli/e2e/tests/build/bundle-budgets.ts:20]: `errorMessage` is an `Error` and this was relying on `errorMessage.toString()` in the `.test` below. When making the `expectToFail` return type stricter this becomes an error.

File: tests/legacy-cli/e2e/utils/process.ts [hunk 2]
```diff
@@ -69,48 +67,65 @@ function _exec(options: ExecOptions, cmd: string, args: string[]): Promise<Proce
...
   const error = new Error();
 
-  // Return log info about the current process status
-  function envDump() {
-    return [
-      `ENV:${JSON.stringify(spawnOptions.env, null, 2)}`,
-      `STDOUT:\n${stdout}`,
-      `STDERR:\n${stderr}`,
-    ].join('\n\n');
-  }
```
Comment (jbedard) [inline tests/legacy-cli/e2e/utils/process.ts:108]: Everything above here was just moved into the `new Promise` so the `waitForMatch` logic in the callbacks can `resolve` the promise.

File: tests/legacy-cli/e2e/utils/process.ts [hunk 3]
```diff
@@ -69,48 +67,65 @@ function _exec(options: ExecOptions, cmd: string, args: string[]): Promise<Proce
...
+        `ENV:${JSON.stringify(spawnOptions.env, null, 2)}`,
+        `STDOUT:\n${stdout}`,
+        `STDERR:\n${stderr}`,
+      ].join('\n\n');
+    }
+
+    childProcess.stdout!.on('data', (data: Buffer) => {
+      stdout += data.toString('utf-8');
+
+      if (options.waitForMatch && stdout.match(options.waitForMatch)) {
```
Comment (jbedard) [inline tests/legacy-cli/e2e/utils/process.ts:93]: Previously this did `data.match` which in theory could fail if the stdout was spread across multiple `'data'` events.

File: tests/legacy-cli/e2e/utils/process.ts [hunk 4]
```diff
@@ -69,48 +67,65 @@ function _exec(options: ExecOptions, cmd: string, args: string[]): Promise<Proce
...
+      }
+
+      data
+        .toString('utf-8')
+        .split(/[\n\r]+/)
+        .filter((line) => line !== '')
+        .forEach((line) => console.error(colors.yellow('  ' + line)));
+    });
+
+    childProcess.on('close', (code: number) => {
```
Comment (jbedard) [inline tests/legacy-cli/e2e/utils/process.ts:128]: Note this changed from `exit` to `close` to ensure stdout/err have completed:

https://nodejs.org/api/child_process.html#event-close
> The 'close' event is emitted after a process has ended and the stdio streams of a child process have been closed

https://nodejs.org/api/child_process.html#event-exit
> When the 'exit' event is triggered, child process stdio streams might still be open.

**Review by alan-agius4 — APPROVED** [review]

**Pull request merged**